File: python/day2.py
""" Solution to Day 2 """

from intcode import Computer
from common import asset


def _part1(program):
    computer = Computer(program)
    computer.run(12, 2)
    print("Part 1:", computer.memory[0])


def _search_noun(computer, start, end, target):
    while end != start + 1:
        mid = (start + end) // 2
        computer.run(mid, 0)
        if computer.memory[0] <= target:
            start = mid
        else:
            end = mid

    return start


def _search_verb(computer, noun, start, end, target):
    while end != start + 1:
        mid = (start + end) // 2
        computer.run(noun, mid)
        if computer.memory[0] <= target:
            start = mid
        else:
            end = mid

    return start


def _part2(program, target):
    computer = Computer(program)
    noun = _search_noun(computer, 0, 100, target)
    verb = _search_verb(computer, noun, 0, 100, target)
    computer.run(noun, verb)
    assert computer.memory[0] == target
    print("Part 2:", noun*100 + verb)


def _main():
    with open(asset("day2.txt")) as file:
        program = [int(value) for value in file.read().split(',')]

    _part1(program)
    _part2(program, 19690720)


if __name__ == "__main__":
    _main()

File: haskell/day2.hs
import Intcode
import Control.Applicative
import System.IO
import System.FilePath
import Debug.Trace

makeTuples = liftA2 (,)

testNounVerb :: [Int] -> (Int, Int) -> (Int, (Int, Int))
testNounVerb program nounVerb = do
    let (noun, verb) = nounVerb
    let a:_:_:program' = program
    let nvProgram = a:noun:verb:program'
    let (Computer memory _ _ _) = runProgram nvProgram
    let result = head (dump memory)
    (result, nounVerb)

part1 :: [Int] -> Int
part1 program = do
    let (result, _) = testNounVerb program (12, 2)
    result

part2 :: [Int] -> Int -> Int
part2 program target = do
    let nounVerbs = makeTuples [0..99] [0..99]
    let results = map (testNounVerb program) nounVerbs
    let (_, (noun, verb)):_ = filter (\(result, _) -> result == target) results
    noun*100 + verb

prep :: Char -> Char
prep char
    | char == ',' = ' '
    | otherwise = char


main = do
    let path = ".." ++ [pathSeparator] ++ "inputs" ++ [pathSeparator] ++ "day2.txt"
    contents <- readFile path
    let values = words (map prep contents)
        program = map read values :: [Int]

    putStrLn ("part 1: " ++ show (part1 program))
    putStrLn ("part 2: " ++ show (part2 program 19690720))
